File: buffer/cpu/src/jvmMain/java/com/wsr/cpu/elementwise/operation/div/JDiv.java
package com.wsr.cpu.elementwise.operation.div;

import java.nio.ByteBuffer;

public class JDiv {
    // 0次元
    public native void divD0ToD1(float x, ByteBuffer y, ByteBuffer result);

    // 1次元
    public native void divD1ToD0(ByteBuffer x, float y, ByteBuffer result);

    public native void divD1ToD1(ByteBuffer x, ByteBuffer y, ByteBuffer result);

    public native void divD1ToD2(ByteBuffer x, ByteBuffer y, int yi, int yj, int axis, ByteBuffer result);

    public native void divD1ToD3(ByteBuffer x, ByteBuffer y, int yi, int yj, int yk, int axis, ByteBuffer result);

    // 2次元
    public native void divD2ToD1(ByteBuffer x, int xi, int xj, ByteBuffer y, int axis, ByteBuffer result);

    public native void divD2ToD3(
            ByteBuffer x,
            int xi,
            int xj,
            ByteBuffer y,
            int yi,
            int yj,
            int yk,
            int axis1,
            int axis2,
            ByteBuffer result
    );

    // 3次元
    public native void divD3ToD1(ByteBuffer x, int xi, int xj, int xk, ByteBuffer y, int axis, ByteBuffer result);

    public native void divD3ToD2(
            ByteBuffer x,
            int xi,
            int xj,
            int xk,
            ByteBuffer y,
            int yi,
            int yj,
            int axis1,
            int axis2,
            ByteBuffer result
    );

    public native void divD3ToD4(
            ByteBuffer x,
            int xi,
            int xj,
            int xk,
            ByteBuffer y,
            int yi,
            int yj,
            int yk,
            int yl,
            int axis1,
            int axis2,
            int axis3,
            ByteBuffer result
    );

    // 4次元
    public native void divD4ToD1(ByteBuffer x, int xi, int xj, int xk, int xl, ByteBuffer y, int axis, ByteBuffer result);

    public native void divD4ToD2(
            ByteBuffer x,
            int xi,
            int xj,
            int xk,
            int xl,
            ByteBuffer y,
            int yi,
            int yj,
            int axis1,
            int axis2,
            ByteBuffer result
    );

    public native void divD4ToD3(
            ByteBuffer x,
            int xi,
            int xj,
            int xk,
            int xl,
            ByteBuffer y,
            int yi,
            int yj,
            int yk,
            int axis1,
            int axis2,
            int axis3,
            ByteBuffer result
    );
}

File: buffer/cpu/src/jvmAndroidMain/kotlin/com/wsr/knist/cpu/elementwise/operation/div/JDiv.kt
package com.wsr.knist.cpu.elementwise.operation.div

object JDiv {
    external fun divD0ToD1(x: Float, y: Long, result: Long)

    external fun divD1ToD0(x: Long, y: Float, result: Long)
    external fun divD1ToD1(x: Long, y: Long, result: Long)
    external fun divD1ToD2(x: Long, y: Long, yi: Int, yj: Int, axis: Int, result: Long)
    external fun divD1ToD3(x: Long, y: Long, yi: Int, yj: Int, yk: Int, axis: Int, result: Long)

    external fun divD2ToD1(x: Long, xi: Int, xj: Int, y: Long, axis: Int, result: Long)
    external fun divD2ToD3(
        x: Long,
        xi: Int,
        xj: Int,
        y: Long,
        yi: Int,
        yj: Int,
        yk: Int,
        axis1: Int,
        axis2: Int,
        result: Long,
    )

    external fun divD3ToD1(x: Long, xi: Int, xj: Int, xk: Int, y: Long, axis: Int, result: Long)
    external fun divD3ToD2(
        x: Long,
        xi: Int,
        xj: Int,
        xk: Int,
        y: Long,
        yi: Int,
        yj: Int,
        axis1: Int,
        axis2: Int,
        result: Long,
    )
    external fun divD3ToD4(
        x: Long,
        xi: Int,
        xj: Int,
        xk: Int,
        y: Long,
        yi: Int,
        yj: Int,
        yk: Int,
        yl: Int,
        axis1: Int,
        axis2: Int,
        axis3: Int,
        result: Long,
    )

    external fun divD4ToD1(x: Long, xi: Int, xj: Int, xk: Int, xl: Int, y: Long, axis: Int, result: Long)
    external fun divD4ToD2(
        x: Long,
        xi: Int,
        xj: Int,
        xk: Int,
        xl: Int,
        y: Long,
        yi: Int,
        yj: Int,
        axis1: Int,
        axis2: Int,
        result: Long,
    )
    external fun divD4ToD3(
        x: Long,
        xi: Int,
        xj: Int,
        xk: Int,
        xl: Int,
        y: Long,
        yi: Int,
        yj: Int,
        yk: Int,
        axis1: Int,
        axis2: Int,
        axis3: Int,
        result: Long,
    )
}
